src/GraphQL/Server.hs

```diff
@@ -164,27 +190,30 @@ lookupValue name args = case find (\(AST.Argument name' _) -> name' == name) arg
 -- the name matches the query. Note that the name is *not* in monad m,
 -- but the value is. This is necessary so we can skip execution if the
 -- name doesn't match.
-data NamedFieldExecutor m = NamedFieldExecutor AST.Name (m GValue.Value)
-
-executeNamedField :: Monad m => AST.Field -> NamedFieldExecutor m -> m (Maybe GValue.ObjectField)
-executeNamedField (AST.Field alias name _ _ _) (NamedFieldExecutor k mValue)
-  | name == k = do
-      value <- mValue
-      let name' = fromMaybe name alias
-      pure (Just (GValue.ObjectField name' value))
-  | otherwise = pure Nothing
-
--- Deconstruct object type signature and handler value at the same
--- time and run type-directed code for each field.
-resolveField :: forall a (m :: Type -> Type). BuildFieldResolver m a => m GValue.ObjectField -> FieldHandler m a -> AST.Selection -> m GValue.ObjectField
-resolveField fallback lh selection@(AST.SelectionField field) = do
-  let fieldExecutor = buildFieldResolver @m @a lh selection
-  case fieldExecutor of
-    Left err -> throwM err
-    Right executor -> do
-      objectField <- executeNamedField field executor
-      maybe fallback pure objectField
-resolveField _ _ f = queryError ("Unexpected Selection value. Is the query normalized?: " <> show f)
+data NamedValueResolver m = NamedValueResolver AST.Name (m (Result GValue.Value))
+
+
+
+-- Iterate through handlers (zipped together with their type
+-- definition) and execute handler if the name matches.
+type ResolveFieldResult = Result (Maybe GValue.ObjectField)
+
+resolveField :: forall a (m :: Type -> Type). BuildFieldResolver m a
+  => FieldHandler m a -> m ResolveFieldResult -> AST.Selection -> m ResolveFieldResult
+resolveField handler nextHandler selection@(AST.SelectionField (AST.Field _ queryFieldName _ _ _)) =
+  case buildFieldResolver @m @a handler selection of
+    -- TODO the fact that this doesn't fit together nicely makes me think that ObjectField is not a good idea)
+    Left err -> pure (Result [err] (Just (GValue.ObjectField queryFieldName GValue.ValueNull)))
+    Right (NamedValueResolver name' resolver) -> runResolver name' resolver
+  where
+    runResolver :: AST.Name -> m (Result GValue.Value) -> m ResolveFieldResult
+    runResolver name' resolver
+      | queryFieldName == name' = do
+          Result errs value <- resolver
+          pure (Result errs (Just (GValue.ObjectField queryFieldName value)))
```

Comment:
The unpacking & reconstruction might perhaps give readability gains, but this could also be written as:

```haskell
Just . GValue.ObjectField queryFieldName <$>  resolver
```